### PR TITLE
20240725-wc_DhAgree_ct

### PR DESCRIPTION
**Pull request by douzzer**
add constant time DH key agreement APIs and related logic:
* `wc_DhAgree_ct()`
* `wolfSSL_DH_compute_key_padded()`

tested with `wolfssl-multi-test.sh ... super-quick-check all-enable-fastmath clang-tidy-intmath intmath-sanitizer`

### CODE REVIEW

File: wolfssl/wolfcrypt/integer.h [hunk 1]
```diff
@@ -311,6 +311,7 @@ MP_API void mp_free (mp_int * a);
 MP_API void mp_forcezero(mp_int * a);
 MP_API int  mp_unsigned_bin_size(const mp_int * a);
 MP_API int  mp_read_unsigned_bin (mp_int * a, const unsigned char *b, int c);
+MP_API int  mp_read_unsigned_bin_ct (mp_int * a, const unsigned char *b, int c);
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/integer.h:314]: integer.c implementation isn't constant time.
Map it to the normal funciton.

File: wolfcrypt/src/dh.c [hunk 2]
```diff
@@ -2040,8 +2040,18 @@ static int wc_DhAgree_Sync(DhKey* key, byte* agree, word32* agreeSz,
...
-        if (ret == 0 && mp_read_unsigned_bin(y, otherPub, pubSz) != MP_OKAY)
-            ret = MP_READ_E;
+        if (ret == 0) {
+            if (ct)
+                ret = mp_read_unsigned_bin_ct(y, otherPub, pubSz);
+            else
+                ret = mp_read_unsigned_bin(y, otherPub, pubSz);
+            if (ret != MP_OKAY)
+                ret = MP_READ_E;
+            #if MP_OKAY != 0
```
Comment (SparkiDev) [inline wolfcrypt/src/dh.c:2050]: Don't need this.

File: wolfcrypt/src/dh.c [hunk 3]
```diff
@@ -2170,11 +2219,17 @@ static int wc_DhAgree_Sync(DhKey* key, byte* agree, word32* agreeSz,
...
-    if (ret == 0 && mp_to_unsigned_bin(z, agree) != MP_OKAY)
-        ret = MP_TO_E;
-
-    if (ret == 0)
-        *agreeSz = (word32)mp_unsigned_bin_size(z);
+    if (ret == 0) {
+        if (ct) {
+            if (mp_to_unsigned_bin_len_ct(z, agree, (int)*agreeSz) != MP_OKAY)
+                ret = MP_TO_E;
+        } else {
```
Comment (SparkiDev) [inline wolfcrypt/src/dh.c:2226]: }
else {

File: src/pk.c [hunk 4]
```diff
@@ -8645,20 +8645,8 @@ int wolfSSL_DH_generate_key(WOLFSSL_DH* dh)
...
- *
- * @param [out] key       Buffer to place shared key.
- * @param [in]  otherPub  Peer's public key.
- * @param [in]  dh        DH key containing private key.
- * @return  -1 on error.
- * @return  Size of shared secret in bytes on success.
- */
-int wolfSSL_DH_compute_key(unsigned char* key, const WOLFSSL_BIGNUM* otherPub,
-    WOLFSSL_DH* dh)
+static int _DH_compute_key_1(unsigned char* key, const WOLFSSL_BIGNUM* otherPub,
```
Comment (SparkiDev) [inline src/pk.c:8648]: Why _1???
_DH_compute_key is a new function.

File: wolfcrypt/src/sp_int.c [hunk 5]
```diff
@@ -17794,17 +17794,7 @@ int sp_unsigned_bin_size(const sp_int* a)
...
- *
- * @param  [out]  a     SP integer.
- * @param  [in]   in    Array of bytes.
- * @param  [in]   inSz  Number of data bytes in array.
- *
- * @return  MP_OKAY on success.
- * @return  MP_VAL when the number is too big to fit in an SP.
- */
-int sp_read_unsigned_bin(sp_int* a, const byte* in, word32 inSz)
+static int _read_unsigned_bin_1(sp_int* a, const byte* in, word32 inSz, int ct)
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:17797]: Remove _1

File: wolfcrypt/src/sp_int.c [hunk 6]
```diff
@@ -17888,12 +17878,43 @@ int sp_read_unsigned_bin(sp_int* a, const byte* in, word32 inSz)
     #endif /* LITTLE_ENDIAN_ORDER */
         }
 #endif
-        sp_clamp_ct(a);
+        if (! ct)
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:17881]: No space after '!'

File: wolfcrypt/src/sp_int.c [hunk 7]
```diff
@@ -17888,12 +17878,43 @@ int sp_read_unsigned_bin(sp_int* a, const byte* in, word32 inSz)
...
+            sp_clamp_ct(a);
```
Comment (SparkiDev) [inline wolfcrypt/src/sp_int.c:17882]: sp_clamp_ct was being called.
Always call it unless there is a real need.
We have to have a clamp.

Don't see a need for the new API if this is changed back.

File: wolfcrypt/src/tfm.c [hunk 8]
```diff
@@ -4040,7 +4040,8 @@ int fp_montgomery_reduce(fp_int *a, fp_int *m, fp_digit mp)
   return fp_montgomery_reduce_ex(a, m, mp, 1);
 }
 
-int fp_read_unsigned_bin(fp_int *a, const unsigned char *b, int c)
+static int _read_unsigned_bin_1(fp_int *a, const unsigned char *b, int c,
```
Comment (SparkiDev) [inline wolfcrypt/src/tfm.c:4043]: Remove _1

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by douzzer** [comment]
retest this please

**Comment by douzzer** [comment]
converted this to a draft because the use case we're trying to accommodate here is for a FIPS wolfCrypt, so the inside-the-boundary changes are problematic at best.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
`#if MP_OKAY != 0` is unneeded. Needs some unit tests.

File: wolfcrypt/src/sp_int.c (hunk 7 above)
Comment (SparkiDev) [reply wolfcrypt/src/sp_int.c:17859]: Ignore this, I misunderstood.

**Review by julek-wolfssl — DISMISSED** [review]

File: wolfssl/wolfcrypt/integer.h (hunk 1 above)
Comment (douzzer) [reply wolfssl/wolfcrypt/integer.h]: fixed

File: wolfcrypt/src/dh.c (hunk 2 above)
Comment (douzzer) [reply wolfcrypt/src/dh.c]: fixed

File: wolfcrypt/src/dh.c (hunk 3 above)
Comment (douzzer) [reply wolfcrypt/src/dh.c]: fixed

File: src/pk.c (hunk 4 above)
Comment (douzzer) [reply src/pk.c]: fixed

File: wolfcrypt/src/sp_int.c (hunk 5 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: fixed

File: wolfcrypt/src/sp_int.c (hunk 6 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: fixed (completely removed the conditionalization of the clamping, as discussed).

File: wolfcrypt/src/sp_int.c (hunk 7 above)
Comment (douzzer) [reply wolfcrypt/src/sp_int.c]: fixed (new API removed).

File: wolfcrypt/src/tfm.c (hunk 8 above)
Comment (douzzer) [reply wolfcrypt/src/tfm.c]: fixed (new API removed).

File: src/pk.c [hunk 9]
```diff
@@ -8773,10 +8761,39 @@ int wolfSSL_DH_compute_key(unsigned char* key, const WOLFSSL_BIGNUM* otherPub,
...
+#endif /* (!HAVE_FIPS || FIPS_VERSION_GE(7,0)) && !HAVE_SELFTEST */
+        {
+            if (wc_DhAgree((DhKey*)dh->internal, key, &keySz, priv,
+                           (word32)privSz, pub, (word32)pubSz) < 0) {
+                WOLFSSL_ERROR_MSG("wc_DhAgree failed");
+                ret = WOLFSSL_FATAL_ERROR;
+            }
+        }
+
+        if ((ret == 0) && ct) {
```
Comment (SparkiDev) [inline src/pk.c:8784]: Why are we padding on constant time?
Didn't wc_DhAgree_ct do that?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/pk.c (hunk 9 above)
Comment (douzzer) [reply src/pk.c:8784]: Yes `wc_DhAgree_ct()` does it, but `wc_DhAgree_ct()` isn't always available, e.g. on older FIPS.  The comment in the code mostly explains this.

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**